File: Homework 6/HashSet.cpp
//
// Created by ozanselte on 12/25/18.
//

#include "HashSet.h"

namespace SELTE_HW6 {
    template<typename E, typename C>
    Iterator<E, C> HashSet<E, C>::iterator() {
        return Iterator<E, C>(container);
    }

    template<typename E, typename C>
    void HashSet<E, C>::add(const E& e) throw(invalid_argument) {
        if(!contains(e)){
            container.insert(container.end(), e);
        }
        else{
            throw invalid_argument("Member is already inside HashSet Collection!");
        }
    }

    template<typename E, typename C>
    void HashSet<E, C>::addAll(Collection<E, C>& c) {
        auto it = c.iterator();
        for(int i = 0; i < c.size(); ++i){
            add(it.step());
        }
    }

    template<typename E, typename C>
    void HashSet<E, C>::clear() {
        container.clear();
    }

    template<typename E, typename C>
    bool HashSet<E, C>::contains(const E& e) const {
        for(auto i : container){
            if(e == i){
                return true;
            }
        }
        return false;
    }

    template<typename E, typename C>
    bool HashSet<E, C>::containsAll(Collection<E, C>& c) const {
        auto it = c.iterator();
        for(int i = 0; i < c.size(); ++i){
            if(!contains(it.step())){
                return false;
            }
        }
        return true;
    }

    template<typename E, typename C>
    bool HashSet<E, C>::isEmpty() const {
        return container.empty();
    }

    template<typename E, typename C>
    void HashSet<E, C>::remove(const E& e) {
        for(auto it = container.begin(); it != container.end(); ++it) {
            if(e == *it){
                container.erase(it);
                return;
            }
        }
    }

    template<typename E, typename C>
    void HashSet<E, C>::removeAll(Collection<E, C>& c) {
        auto it = c.iterator();
        for(int i = 0; i < c.size(); ++i){
            remove(it.step());
        }
    }

    template<typename E, typename C>
    HashSet<E, C>& HashSet<E, C>::retainAll(Collection<E, C>& c) {
        static HashSet<E, C> temp;
        temp.clear();
        auto it = c.iterator();
        for(int i = 0; i < c.size(); i++){
            E e = it.step();
            if(contains(e)){
                temp.add(e);
            }
        }
        return temp;
    }

    template<typename E, typename C>
    size_t HashSet<E, C>::size() const {
        return container.size();
    }
}

File: Homework 8/com/ozanselte/HashSet.java
package com.ozanselte;

/**
 * Implementation of Set interface functions
 *
 * @author Ozan Şelte
 * @version 1.0.0
 * @since 2019-01-11
 * @see com.ozanselte.Set
 * @param <E> <code>Integer</code> or <code>String</code>
 */
public class HashSet<E> implements Set<E> {

    private E[] arr;

    public Iterator<E> iterator() {
        return new Iterator<E>(arr);
    }

    /**
     * Ensures that this collection contains the specified element.
     * @param e The element
     * @throws UnsupportedOperationException in case of passing already contained element.
     */
    @SuppressWarnings("unchecked")
    public void add(E e) throws UnsupportedOperationException {
        if (this.contains(e)) {
            throw new UnsupportedOperationException("Member is already inside HashSet Collection!");
        }
        else {
            E[] futureArr = (E[])new Object[this.size() + 1];
            if (this.size() > 0) {
                System.arraycopy(this.arr, 0, futureArr, 0, this.size());
            }
            futureArr[this.size()] = e;
            arr = futureArr;
        }
    }

    public void addAll(Collection<E> c) {
        Iterator<E> it = c.iterator();
        for (int i = 0; i < c.size(); i++) {
            this.add(it.step());
        }
    }

    @SuppressWarnings("unchecked")
    public void clear() {
        this.arr = (E[])new Object[0];
    }

    public boolean contains(E e) {
        for (int i = 0; i < this.size(); i++) {
            if (this.arr[i].equals(e)) {
                return true;
            }
        }
        return false;
    }

    public boolean containsAll(Collection<E> c) {
        Iterator<E> it = c.iterator();
        for (int i = 0; i < c.size(); i++) {
            if (!this.contains(it.step())){
                return false;
            }
        }
        return true;
    }

    public boolean isEmpty() {
        return this.size() <= 0;
    }

    @SuppressWarnings("unchecked")
    public void remove(E e) {
        for (int i = 0; i < this.size(); i++) {
            if (this.arr[i].equals(e)) {
                E[] futureArr = (E[])new Object[arr.length - 1];
                if (0 < i) {
                    System.arraycopy(arr, 0, futureArr, 0, i);
                }
                if (futureArr.length > i) {
                    System.arraycopy(arr, i+1, futureArr, i, arr.length-i-1);
                }
                arr = futureArr;
                return;
            }
        }
    }

    public void removeAll(Collection<E> c) {
        Iterator<E> it = c.iterator();
        for (int i = 0; i < c.size(); i++) {
            this.remove(it.step());
        }
    }

    public HashSet<E> retainAll(Collection<E> c) {
        HashSet<E> temp = new HashSet<E>();
        Iterator<E> it = c.iterator();
        for (int i = 0; i < c.size(); i++) {
            E e = it.step();
            if (this.contains(e)) {
                temp.add(e);
            }
        }
        return temp;
    }

    public int size() {
        if (null == arr) {
            return 0;
        }
        return arr.length;
    }
}
